.github/workflows/atlas-tests.yml

```diff
@@ -0,0 +1,134 @@
+            ~/.cargo/registry/index/
+            ~/.cargo/registry/cache/
+            ~/.cargo/git/db/
+          # key: ${{ github.event.repository.name }}-cargo
+          key: ${{ github.event.repository.name }}-${{ github.sha }}-cargo
+
+      ## Restore bitcoin cache data
+      - name: Restore Bitcoin Binary Cache
+        id: restore_bitcoin_cache
+        uses: actions/cache/restore@88522ab9f39a2ea568f7027eddc7d8d8bc9d59c8 # v3.3.1
+        with:
+          fail-on-cache-miss: true
+          path: /tmp/bitcoin
+          key: ${{ github.event.repository.name }}-bitcoin-binaries
+
+      ## Restore nextest archive cache data
+      - name: Restore Test Archive Cache
+        id: restore_test-archive_cache
+        uses: actions/cache/restore@88522ab9f39a2ea568f7027eddc7d8d8bc9d59c8 # v3.3.1
+        with:
+          fail-on-cache-miss: true
+          path: /tmp/test_archive.tar.zst
+          # key: ${{ github.event.repository.name }}-test-archive
+          key: ${{ github.event.repository.name }}-${{ github.sha }}-test-archive
+
+      ## Restore nextest target cache data
+      - name: Restore Target Cache
+        id: restore_target_cache
+        uses: actions/cache/restore@88522ab9f39a2ea568f7027eddc7d8d8bc9d59c8 # v3.3.1
+        with:
+          fail-on-cache-miss: true
+          path: ./target
+          # key: ${{ github.event.repository.name }}-target
+          key: ${{ github.event.repository.name }}-${{ github.sha }}-target
+
+      ## Symlink bitoind to $PATH
+      - name: Link Bitcoin Binary
+        id: bitcoin_binary
+        run: |
+          sudo ln -s /tmp/bitcoin/bin/bitcoind /bin/
+          ls -al /bin/bitcoind
+
+      ## Run test matrix using cached archive file
+      - name: Run Tests
+        id: run_tests
+        timeout-minutes: 30
+        run: |
+          cargo nextest run \
+            --test-threads num-cpus \
+            --run-ignored all \
+            --retries 2 \
+            --final-status-level fail \
+            --fail-fast \
+            --archive-file /tmp/test_archive.tar.zst \
+            "${{ matrix.test-name }}"
+
+      ## Generate code coverage using grcov
+      - name: Run grcov
+        id: run_grcov
+        run: |
+          grcov . --binary-path ./target/debug/ -s . -t lcov --branch --ignore-not-existing --ignore "/*" -o lcov.info || exit 1
+          echo "created=true\n" >> $GITHUB_OUTPUT
+
+      ## Upload coverage to codecov
+      ##   - use wretry action, to retry on a failed upload 3x
+      - name: Upload Code Coverage
+        id: code_coverage
+        uses: Wandalen/wretry.action@a163f62ae554a8f3cbe27b23db15b60c0ae2e93c # v1.3.0
+        with:
+          action: codecov/codecov-action@eaaf4bedf32dbdc6b720b63067d99c4d77d6047d # v3.1.4
+          with: |
+            files: ./lcov.info
+            name: ${{ matrix.test-name }}
+            fail_ci_if_error: false
+            verbose: true
+          attempt_limit: 3
+          attempt_delay: 10000
```

Comment:
Do we think it could be clearer if we merged the two ./lcov.info files generated from these tests and sent them up to codecov in one go? In my mind we don't care where the coverage came from but I don't know if codecov merges the two coverage results if we upload two for the same commit under different names.

Reply:
that i'm unsure of - this is the way the CI currently works, the only changes i've added here (outside of  moving the atlas tests to a new workflow) are retrying the upload (we've seen spurious 500 errors from codecov), and locking the external workflow to a specific commit has (vs using a version/master)

Reply:
i suppose it's possible to upload them all at once - off the top of my head, we could save the output from each matrix job as an artifact, then a new job to upload the artifacts all at once